src/Bijectors.jl

```diff
@@ -180,8 +180,10 @@ function link(
     x::AbstractVecOrMat{<:Real},
     ::Type{Val{proj}} = Val{true}
 ) where {proj}
-    return SimplexBijector{proj}()(x)
+    return _bcall(SimplexBijector{proj}(), x)
 end
+# Workaround because ::Type{Val{proj}} seems to break Zygote adjoint
+_bcall(f, x) = f(x)
```

Comment:
What do you mean?

Reply:
I just read your comment about `p::Type{Val{proj}}` breaking Zygote's adjoint, and was wondering if that could be fixed in another way, by, e.g., explicitly using the pullback of a closure over `p`. But I'm not sure if that makes sense even :smile:

Reply:
Question: do we really need the `Type{Val{proj}}` now that the `link` and `invlink` methods for `SimplexDistribution` are simply calling `SimplexBijector` which itself just extracts the actual value `proj`?

Reply:
We can still pass `proj` to `link` or `invlink` which is used in the definition of the bijector, e.g:
```julia
function link(
    d::SimplexDistribution,
    x::AbstractVecOrMat{<:Real},
    ::Type{Val{proj}} = Val{true}
) where {proj}
    return _bcall(SimplexBijector{proj}(), x)
end
```

Reply:
But why not just make it
```julia
function link(
    d::SimplexDistribution,
    x::AbstractVecOrMat{<:Real},
    proj::Bool = true
) where {proj}
    return _bcall(SimplexBijector{proj}(), x)
end
```
?
I believe the reasoning was that constant-propagation didn't always work as well as `Type{Val}`. But now `SimplexBijector` doesn't work with `Type{Val}` anymore and so aren't we now reliant on constant-propagation working for the `SimplexBijector{proj}()` call anyways?

Reply:
Hmm currently `proj` is type information so we don't rely on run-time constant propagation. `proj` is a compile-time constant. Using your suggestion would be depending on the run-time constant propagation which is fine if it works and in this simple case, it might actually work fine. I don't mind using the simpler version if you do benchmarks and find they are identical, but maybe in another PR.

Reply:
Okay, so this might just be me not properly understanding how `Val` works.

Say I do something like:
```julia
f(x::Val{x}) where {x} = g(x)
g(x) = h(x)
# 1000s of nested functions to the point where the compiler doesn't inline them all automatically
# ...
z(x) = x
```
Is `x` inside scope of `z(x)` known at compile-time?

> I don't mind using the simpler version if you do benchmarks and find they are identical, but maybe in another PR.

Sure, but in this PR you do a lot of work to circumvent the issue of Zygote being incompatible with `Type{Val{proj}}`, right? So seems like it would be great to clear that up before we make this change? :confused:

Reply:
> EDIT: Btw found this: https://discourse.julialang.org/t/getproperty-decorations-inheritance-in-0-7/11237/9 Seems we can drop the Val without issues, no?

Note though Kristoffer's comment further down in the same issue: https://discourse.julialang.org/t/getproperty-decorations-inheritance-in-0-7/11237/14

Reply:
@torfjelde we can drop the `Val` yes if constant propagation works properly, which it should. For your examples, only the type of `x` in `z(x)` is known at compile-time not its value. However, what constant propagation does is if say we have `h() = z(1)`, then the constant 1 will propagate inside the function `z` and the function `h` will be compiled to:
```julia
julia> @code_typed h()
CodeInfo(
1 ─     return 1
) => Int64
```
This is known as inter-procedural constant propagation. So inline and compile-time constants can get propagated through function barriers compiling more efficient methods knowing that the value of the runtime variable `x` is actually constant. Here is an example where a compile-time constant gets propagated:
```julia
julia> h(::Val{x}) where {x} = z(x)
h (generic function with 2 methods)

julia> @code_typed h(Val(1))
CodeInfo(
1 ─     return 1
) => Int64
```

Reply:
> Sure, but in this PR you do a lot of work to circumvent the issue of Zygote being incompatible with Type{Val{proj}}, right?

Fair enough, I will do it in this PR then.

Reply:
> Note though Kristoffer's comment further down in the same issue: https://discourse.julialang.org/t/getproperty-decorations-inheritance-in-0-7/11237/14

With that comment + what you just said @mohamed82008, it finally makes sense! Thank you both:) I now see why it's useful in this case.

> Fair enough, I will do it in this PR then.

Given that I finally understood why we bother, ehhmm I'm leaving the decision to you :+1: